rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,588 @@
+
+```python
+class CustomFieldsSchema(Schema):
+     """Marshmallow schema for custom fields.
+
+     Loads all schemas from the configured fields.
+     Uses the singleton pattern to avoid loading multiple times.
+     """
+
+     def __init__(self, *args, **kwargs):
+         """constructor."""
+         super().__init__(*args, **kwargs)
+         config = current_app.config.get("RDM_CUSTOM_FIELDS", {})
+         self.fields = {
+             field_key: getattr(field, "schema")()
+             for field_key, field in config.items()
+         }
+         self._schema = Schema.from_dict(self.fields)()
+
+     def _serialize(self, obj, **kwargs):
+         """Dumps the custom fields values."""
+         return self._schema.dump(obj)
+
+     def _deserialize(self, data, **kwargs):
+         """Loads the custom fields values."""
+         return self._schema.load(data=data)
+```
+
+#### Linking vocabulary custom fields to a record
+
+> Note: in this sub section the term _record_ is treated as a generic term, namely a document. It will refer to bibliographic and community records as _records_.
+
+Simple custom fields types are added to the record's data model and dumped in its json blob. However, the vocabulary type works differently as it is a relation to another record, and it needs to be linked. Meaning, the selected value need to be checked for existence when a record is created and dereferenced when it is read.
+
+The programmatic API to access records' relations is kept intact. For example, a field named `myrelationfield` would be accessed as `record.relations.myrelationfield`. However, a new class `MultiRelationsField` has been implemented to accept not only implementations of `RelationBase` but also `RelationsField`.
+
+`CustomFieldsRelation` is an implementation of `RelationsField` that injects vocabulary custom fields in the record's relations attribute. In addition, to avoid name collisions these new fields are prefixed with `custom_`.
+
+For example, the previously configured `experiments` custom field is accessed as `record.relations.custom_experiments`, while `record.relations.custom` would give an attribute error.
+
+```python
+class Record:
+    relations = MultiRelationsField(
+        custom_fields=CustomFieldsRelation(),
+    )
+```
+
+The dereferencing part is transparent, i.e. no code needs to be altered since the API is still the same. It is carried out by the `RelationsComponent` which will dereference the relations.
+
+### Creating a _custom field_
+
+Once the custom fields are configured in the `invenio.cfg` file, they need to be added to the record's mapping, so their values are properly indexed in Elasticsearch. That can be done using the CLI.
+
+**create**
+
+You can add fields to the records and community data models using the following command. You can pass any number of configured fields.
+
+```shell
+invenio-cli custom-fields create [-f <fieldname>] [-f <fieldname>]
```

Comment:
Question: Can this not be part of the ``index init``? I.e. if I configure the custom fields in ``invenio.cfg`` before creating the index, then it should just be there from the beginning and not having to run this command. 

Comment: Specify here that the problem being solved, is to update the mapping once it has already been created (i.e. you add a custom field after running the instance)

Reply:
Discussion:
- open issue
- two use cases:
    - `invenio-cli services setup` for new instances
    - `invenio-cli upgrade` since it could be considered a migration - schema upgrade

Reply:
https://github.com/inveniosoftware/invenio-rdm-records/issues/1007